### PR TITLE
cli: Support nodelocal file upload in CLI

### PR DESCRIPTION
**Pull request by g3orgia**
This change uses the changes made in #42748, which enabled
COPY to upload files. We are now implementing the CLI to include
a new command: `nodelocal upload`. This command takes in a
source file to upload, and a destination filename. It will
then use the COPY command to upload the file and drop it
on the gateway node's local file system, at:

`externalIODir/destination/filename`

This PR also includes the CLI testing for this feature as
well as a bug fix for the previous PR #42748:
- If writing the file fails, we do not delete the file
 that was created. This PR fixes that.

Release note (cli change): Adds a nodelocal command
that can be used to upload file:
`cockroach nodelocal upload location/of/file destination/of/file`

### CODE REVIEW

**Comment by g3orgia** [comment]
There's two things that still need to be changed in this PR:
- Verify the version (which is commented out right now). I'm unsure which version I should be verifying. This needs to be changed before this PR gets merged.
- Return the nodelocal URI of the file that was written. Is there a way of figuring out the gateway node's `nodeID` from the CLI? If not, I may need to make a separate PR to pass that information through.

File: pkg/sql/copy_file_upload_test.go
Comment (mjibson) [inline pkg/sql/copy_file_upload_test.go]: The `CREATE DATABASE` above can be removed now.

File: pkg/sql/copy_file_upload.go
Comment (mjibson) [inline pkg/sql/copy_file_upload.go]: What about if f.writeError has an error? Can/should the API be changed so there's only 1 possible error produced?

File: pkg/sql/conn_executor.go
Comment (mjibson) [inline pkg/sql/conn_executor.go]: merge this line into the if:
`if table := cmd.Stmt.Table; table.Table() == fileUploadTable ...`

File: pkg/cli/nodelocal.go
Comment (mjibson) [inline pkg/cli/nodelocal.go]: remove

File: pkg/cli/nodelocal.go
Comment (mjibson) [inline pkg/cli/nodelocal.go]: I'd just ignore this version check. People using out-of-sync client and server will sometimes have problems and that's ok.

File: pkg/cli/nodelocal.go
Comment (mjibson) [inline pkg/cli/nodelocal.go]: Remove this call and just use `f.Stat()`. Otherwise there's a race condition since the file could change between the call to stat and the call to open.

File: pkg/cli/nodelocal.go
Comment (mjibson) [inline pkg/cli/nodelocal.go]: n can be > 0 even with an EOF, so you still need to send the number of bytes read. i'm surprised the tests passed.

Ah, it's because the API for io.Reader is that n can be > 0 even with EOF. But os.File.Read specifies 0, EOF at the end. You have two choices:

1) Teach the code here to always deal with n > 0. You should do this (as per the io.Reader API) only checking for error if n == 0, since that API says any error should be returned during subsequent calls. This means you don't have to do any weird error checking after processing the data, which is nice.
2) Have this function take an os.File, which guarantees that 0, EOF is returned on EOF.

Both of these are fine and you could do whichever you like more. I'd probably do the first since it feels safer and is also more testable since it doesn't depend on a file coming in.

File: pkg/cli/nodelocal.go
Comment (mjibson) [inline pkg/cli/nodelocal.go]: can you just exec without casting to a string? that'd save a memory copy.

File: pkg/cli/nodelocal.go
Comment (mjibson) [inline pkg/cli/nodelocal.go]: since you don't ever close the reader, just require an `io.Reader` here

File: pkg/cli/nodelocal.go
Comment (mjibson) [inline pkg/cli/nodelocal.go]: need a reader.Close() somewhere here

File: pkg/cli/nodelocal.go
Comment (mjibson) [inline pkg/cli/nodelocal.go]: I don't think you need to reallocate this slice for each iteration.

File: pkg/cli/nodelocal.go
Comment (g3orgia) [reply pkg/cli/nodelocal.go]: When I exec as just bytes, it prints the string representation of the bytes to stdin, which gets received that way on the other side. I'm unsure how to receive that on the other side, and convert them back to strings.

File: pkg/cli/nodelocal.go
Comment (g3orgia) [reply pkg/cli/nodelocal.go]: Yes, I will do the first. It would cause less confusion if we ever used any other implementation of reader.

File: pkg/sql/copy_file_upload.go
Comment (g3orgia) [reply pkg/sql/copy_file_upload.go]: I will think about restructuring this. However, since this is on two sides of an `io.Pipe`, we still need to capture either side's failures.

File: pkg/cli/nodelocal.go
Comment (mjibson) [reply pkg/cli/nodelocal.go]: Have you tried `stmt.Exec(send[:n])`? I thought that's how lib/pq's Copy was supposed to work.

File: pkg/cli/nodelocal.go
Comment (mjibson) [inline pkg/cli/nodelocal.go]: In order to make it obvious that the n > 0 case is handled correctly, I would rejigger this to:

```
if n > 0 {
	_, err := stmt.Exec([]driver.Value{string(send[:n])})
	if err != nil {
		return err
	}
} else if err == io.EOF {
	break
} else if err != nil {
	return err
}
```

File: pkg/cli/nodelocal.go
Comment (g3orgia) [reply pkg/cli/nodelocal.go]: Hmm doesn't seem like it. I'll look into how we can structure this better, make use of some of the better suited APIs.

**Review by madelynnblue — APPROVED** [review]

**Comment by g3orgia** [comment]
Two changes:
- Outputs nodelocal URL of uploaded file
- Error handling using `io.Pipe` instead of creating our own channel

File: pkg/sql/copy_file_upload.go
Comment (mjibson) [inline pkg/sql/copy_file_upload.go]: this defer doesn't do anything extra. you can just put the logic right in the function.

File: pkg/sql/copy_file_upload.go
Comment (mjibson) [inline pkg/sql/copy_file_upload.go]: this can be a simple `sync.WaitGroup` at this point since it's not exchanging any information over the chan, and is just a synchronization point.

File: pkg/cli/nodelocal.go
Comment (mjibson) [inline pkg/cli/nodelocal.go]: missing `:` after `nodelocal`

**Review by madelynnblue — APPROVED** [review]

**Pull request merged**
